contracts/fio.perms/fio.perms.cpp

```diff
@@ -0,0 +1,477 @@
+            appConfig = configsSingleton.get_or_default(config());
+        }
+
+
+        /*
+         * This action will check if a permission exists for the specified arguments, if it does not
+         * yet exist in the permissions table a new record will be added, the accesses table will also
+         * be updated to indicate the grantee account access that has been granted. please see the code
+         * for error logic and parameter descriptions.
+         */
+        [[eosio::action]]
+        void
+        addperm(const name &grantee_account,
+                const string &permission_name,   //one permission is permitted register_address_on_domain
+                const string &permission_info,   //this is empty for FIP-40, an extensibility field for the future.
+                const string &object_name,       //the name of the fio domain
+                const int64_t &max_fee,
+                const string &tpid,
+                const name &actor
+                ) {
+
+
+            print("addperm --      called. \n");
+
+            require_auth(actor);
+            string useperm = makeLowerCase(permission_name);
+
+
+            fio_400_assert(permission_name.length()  > 0, "permission_name", permission_name,
+                           "Permission name is invalid", ErrorInvalidPermissionName);
+            //  error if permission name is not the expected name register_address_on_domain.
+            //one permission name is integrated for fip 40, modify this logic for any new permission names
+            //being supported
+            fio_400_assert(useperm.compare(REGISTER_ADDRESS_ON_DOMAIN_PERMISSION_NAME) == 0, "permission_name", permission_name,
+                           "Permission name is invalid", ErrorInvalidPermissionName);
+
+            // error if permission info is not empty.
+            fio_400_assert(permission_info.size()  == 0, "permission_info", permission_info,
+                           "Permission info is invalid", ErrorInvalidPermissionInfo);
+            // error if object name is not * or is not in the domains table
+            fio_400_assert(object_name.size()  > 0, "object_name", object_name,
+                           "Object name is invalid", ErrorInvalidObjectName);
+
+            //verify domain name, and that domain is owned by the actor account.
+            FioAddress fa;
+            getFioAddressStruct(object_name, fa);
+
+            fio_400_assert(fa.domainOnly, "domonlyobject_name", object_name, "Invalid object name",
+                           ErrorInvalidObjectName);
+
+            const uint128_t domainHash = string_to_uint128_hash(fa.fiodomain.c_str());
+            auto domainsbyname = domains.get_index<"byname"_n>();
+            auto domains_iter = domainsbyname.find(domainHash);
+
+            fio_400_assert(domains_iter != domainsbyname.end(), "domnotfound object_name", object_name,
+                           "Invalid object name",
```

Comment:
Since this error is specific to the domain object, I recommend we use a clearer error like "FIO Domain not found"